src/test/unit/althold_unittest.cc

```diff
@@ -0,0 +1,163 @@
+    PG_REGISTER(altholdConfig_t, altholdConfig, PG_ALTHOLD_CONFIG, 0);
+
+    extern altHoldState_t altHoldState;
+    void altHoldReset(void);
+    void altHoldProcessTransitions(void);
+    void altHoldInit(void);
+    void altHoldUpdate(void);
+    // end why ??
+}
+
+#include "unittest_macros.h"
+#include "gtest/gtest.h"
+
+uint32_t millisRW;
+uint32_t millis() {
+    return millisRW;
+}
+
+TEST(AltholdUnittest, altHoldTransitionsTest)
+{
+    altHoldInit();
+    EXPECT_EQ(altHoldState.isAltHoldActive, false);
+
+    flightModeFlags |= ALTHOLD_MODE;
+    millisRW = 42;
+    altHoldUpdate();
+    EXPECT_EQ(altHoldState.isAltHoldActive, true);
+
+    millisRW = 56;
+    flightModeFlags ^= ALTHOLD_MODE;
+    altHoldUpdate();
+    EXPECT_EQ(altHoldState.isAltHoldActive, false);
+
+    flightModeFlags |= ALTHOLD_MODE;
+    millisRW = 64;
+    altHoldUpdate();
+    EXPECT_EQ(altHoldState.isAltHoldActive, true);
+
+//     millisRW = 64 + 0.3f * ALTHOLD_ENTER_PERIOD;
+//     altHoldUpdate();
+// 
+//     millisRW = 64 + 0.9f * ALTHOLD_ENTER_PERIOD;
+//     altHoldUpdate();
+// 
+//     millisRW = 64 + 1.4f * ALTHOLD_ENTER_PERIOD;
+//     altHoldUpdate();
+// 
+//     millisRW = 10042;
+//     flightModeFlags ^= ALTHOLD_MODE;
+//     altHoldUpdate();
+// 
+//     millisRW = 10042 + 0.3f * ALTHOLD_MAX_ENTER_PERIOD;
+//     altHoldUpdate();
+// 
+//     millisRW = 10042 + 0.5f * ALTHOLD_MAX_ENTER_PERIOD;
+//     altHoldState.throttleOut = 0.5f;
+//     altHoldUpdate();
+}
+
+TEST(AltholdUnittest, altHoldTransitionsTestUnfinishedExitEnter)
+{
+    altHoldInit();
+    EXPECT_EQ(altHoldState.isAltHoldActive, false);
+
+    flightModeFlags |= ALTHOLD_MODE;
+    millisRW = 42;
+    altHoldUpdate();
+    EXPECT_EQ(altHoldState.isAltHoldActive, true);
+//     EXPECT_EQ(altHoldState.enterTime, 42);
+//     EXPECT_EQ(altHoldState.exitTime, 0);
+// 
+//     millisRW += ALTHOLD_ENTER_PERIOD * 2;
+//     altHoldUpdate();
+// 
+//     flightModeFlags ^= ALTHOLD_MODE;
+//     altHoldUpdate();
+//     millisRW += 0.5f * ALTHOLD_MAX_ENTER_PERIOD;
+//     altHoldUpdate();
+// 
+//     flightModeFlags |= ALTHOLD_MODE;
+//     millisRW += 1;
+//     altHoldUpdate();
+}
+
+// STUBS
+
+extern "C" {
+    acc_t acc;
+
+    void pt2FilterInit(pt2Filter_t *altHoldDeltaLpf, float) {
+        UNUSED(altHoldDeltaLpf);
+    }
+    float pt2FilterGain(float, float) {
+        return 0.0f;
+    }
+    float pt2FilterApply(pt2Filter_t *altHoldDeltaLpf, float) {
+        UNUSED(altHoldDeltaLpf);
+        return 0.0f;
+    }
+
+
+    float getAltitude(void) { return 0.0f; }
+    float rcCommand[4];
+
+    float getRcDeflection(int)
+    {
+        return 0;
+    }
+
+    void parseRcChannels(const char *input, rxConfig_t *rxConfig)
+    {
+        UNUSED(input);
+        UNUSED(rxConfig);
+    }
+
+    int16_t debug[DEBUG16_VALUE_COUNT];
+    uint8_t debugMode;
+
+    uint8_t armingFlags = 0;
+    uint8_t stateFlags = 0;
+    uint16_t flightModeFlags = 0;
+}
```

Comment:
Missing return on EOF
```suggestion
}

```